Added linalg.cholesky

IvanYashchuk opened the pull request:
This PR adds `torch.linalg.cholesky` function that matches `numpy.linalg.cholesky`.

Fixed `lda` argument to `lapackCholesky` calls.
Added `random_hermitian_pd_matrix` helper function for tests.

Ref #42666.

IvanYashchuk committed:
wip linalg.cholesky

IvanYashchuk committed:
Added xfailed test case

IvanYashchuk committed:
Added cholesky to csrc/api/include/torch/linalg.h

IvanYashchuk committed:
Updated example in docs

IvanYashchuk committed:
Added random_hermitian_pd_matrix for the test

IvanYashchuk committed:
Use random_hermitian_pd_matrix in the test_torch/cholesky

IvanYashchuk commented:
I'm not sure whether what I did in `torch/csrc/api/include/torch/linalg.h` is correct.
Existing functions have names `linalg_det`, `linalg_norm`, and there are in `linalg` namespace. Are they accessed with `torch::linalg::linalg_det`?

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into linalg-cholesky

kurtamohler commented on test/test_linalg.py:
```diff
@@ -921,6 +922,43 @@ def test_nuclear_norm_exceptions_old(self, device):
+        from torch.testing._internal.common_utils import random_hermitian_pd_matrix
+        A = random_hermitian_pd_matrix(shape, *batch, dtype=dtype, device=device)
+        expected_L = np.linalg.cholesky(A.cpu().numpy())
+        actual_L = torch.linalg.cholesky(A)
```
Does this produce the aforementioned RuntimeError? If so, then I think it's probably better to use a `with self.assertRaisesRegex(...):`, rather than `expectedFailure`.

IvanYashchuk replied:
The idea to use an xfail test (`unittest.expectedFailure`) because that is a bug we can't fix right now.
I would expect to use `self.assertRaisesRegex(...)` (and consequently mark the test as 'pass') for the intended errors that are not bugs.

kurtamohler commented on test/test_torch.py:
```diff
@@ -7761,19 +7761,9 @@ def cholesky_test_helper(n, batch_dims, upper):
+        from torch.testing._internal.common_utils import random_hermitian_pd_matrix
 
-        # This is a workaround while there is no support for complex random_symmetric_pd_matrix
-        if dtype.is_complex:
```
Is this being removed because you added support for complex random_symmetric_pd_matrix (I'm not sure what that is)? Is this because of your `lda` fix?

IvanYashchuk replied:
Cholesky requires the input to be Hermitian positive definite, which is symmetric positive definite for the real-valued case. This is removed because I added `random_hermitian_pd_matrix`.
For real dtypes `random_hermitian_pd_matrix` is equivalent to `random_symmetric_pd_matrix` since `.conj()` does nothing.

kurtamohler commented on test/test_linalg.py:
```diff
@@ -921,6 +922,43 @@ def test_nuclear_norm_exceptions_old(self, device):
+
+        shapes = (0, 3, 5)
+        batches = ((), (3, ), (2, 2))
+        for shape, batch in itertools.product(shapes, batches):
```
Maybe it would be good to try at least one large input, with a few thousand elements. You could convert this product into a list and then append an extra case.

IvanYashchuk replied:
Sure, I will add it.

kurtamohler commented on torch/linalg/__init__.py:
```diff
@@ -8,6 +8,44 @@
+
+Returns the Cholesky decomposition.
+
+Computes the Cholesky decomposition of a Hermitian (or symmetric for real-valued matrices)
```
If these input requirements are not met, do we throw an error? Seems like numpy does:

```
>>> numpy.linalg.cholesky(numpy.random.randn(10, 10))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<__array_function__ internals>", line 6, in cholesky
  File "/home/endoplasm/miniconda3/lib/python3.7/site-packages/numpy/linalg/linalg.py", line 764, in cholesky
    r = gufunc(a, signature=signature, extobj=extobj)
  File "/home/endoplasm/miniconda3/lib/python3.7/site-packages/numpy/linalg/linalg.py", line 91, in _raise_linalgerror_nonposdef
    raise LinAlgError("Matrix is not positive definite")
numpy.linalg.LinAlgError: Matrix is not positive definite
```

I think we should probably have some `self.assertRaisesRegex` tests for any input error cases we expect. Probably would also be good to make sure numpy gives errors for the same cases.

kurtamohler replied:
Also I think it would be nice to mention in the docs that an error will be thrown if the input requirements are not met.

IvanYashchuk replied:
Yes, the error is thrown if the input is not valid, I've added `assertRaisesRegex` checks.
I also added a note to the docs:
> If the :attr:`input` is not Hermitian positive-definite matrix a RuntimeError is raised
          saying that the input is singular and mentioning which minor of the input matrix is not positive-definite.

kurtamohler commented on test/test_linalg.py:
```diff
@@ -921,6 +922,43 @@ def test_nuclear_norm_exceptions_old(self, device):
 
+    @skipCUDAIfNoMagma
+    @skipCPUIfNoLapack
+    @unittest.skipIf(not TEST_NUMPY, "NumPy not found")
```
numpy is always imported now.

kurtamohler reviewed (changes requested):
Looks good overall to me.

Does `linalg.cholesky` support JIT and autograd? If so, maybe we should have a test case for it in `test/test_linalg.py:test_autograd_and_jit`

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into linalg-cholesky

IvanYashchuk committed:
No need for skip if numpy not found anymore

IvanYashchuk committed:
Added larger input case

IvanYashchuk committed:
Added assertRaises tests for cholesky

Added precisionOverride. MAGMA cholesky seem to disagree a bit with NumPy on large input hence 1e-2 atol

IvanYashchuk committed:
Added a note to the docs that the error is given if the input is not
positive definite

IvanYashchuk committed:
Enabled autograd for linalg_cholesky

IvanYashchuk committed:
Added a note to the documentation about complex support

IvanYashchuk commented:
`linalg.cholesky` now supports autograd. And I added the test for that. It currently works only for CPU for complex inputs though, because `triangular_solve_cuda` is not implemented yet for complex numbers.

I don't know what is needed to be done for the JIT support.

kurtamohler commented on test/test_linalg.py:
```diff
@@ -922,6 +923,90 @@ def test_nuclear_norm_exceptions_old(self, device):
+        actual_L = torch.linalg.cholesky(A)
+        self.assertEqual(actual_L, expected_L)
+
+    # TODO: enable CUDA tests once
```
Maybe would also be good to mention that once jit and CUDA complex support are added, this test should be moved to `test_autograd_and_jit`. Or if the tests in `test_autograd_and_jit` have been moved into `common_methods_invocations.py`, it should be moved there.

Jit and CUDA complex support can be done in future PRs.

IvanYashchuk replied:
The test for `torch.cholesky`from `test/test_autograd.py` is not part of `common_methods_invocations.py:method_tests`. Maybe the reason is that it requires making the input hermitian for gradcheck to work correctly
```python
def func(x):
    x = 0.5 * (x + x.transpose(-1, -2).conj()) # Make `x` Hermitian
    return torch.linalg.cholesky(x)
```
`func(x)` is being tested not `torch.linalg.cholesky(x)` directly. So I don't know whether it can be moved later to `test_autograd_and_jit`/`common_methods_invocations.py`.

kurtamohler replied:
As far as I understand, I don't think that prevents us from putting this in `common_methods_invocations.py:method_tests`. The second element of an entry in `method_tests` can be a contructing function, for instance: https://github.com/pytorch/pytorch/blob/master/torch/testing/_internal/common_methods_invocations.py#L521

So I think it would be possible to make a constructing function for `cholesky` too.

mruberry commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -8284,6 +8284,11 @@
 #
 # See linalg_det as an example.
 
+- func: linalg_cholesky(Tensor self) -> Tensor
```
What's the plan for the out variant of this function?

IvanYashchuk replied:
I added it now.

mruberry commented on test/test_linalg.py:
```diff
@@ -922,6 +923,90 @@ def test_nuclear_norm_exceptions_old(self, device):
         self.assertRaisesRegex(IndexError, "Dimension out of range", torch.norm, x, "nuc", (0, 2))
 
 
+    @precisionOverride({torch.float: 1e-2, torch.cfloat: 1e-4})
```
The absolute tolerance for float32 is surprisingly large. What's going on there?

IvanYashchuk replied:
I assume that NumPy could silently convert the input to fp64 and then back to fp32.
CPU tests pass for `1e-3` tolerance. GPU passes only with `1e-2`. It doesn't mean that the decomposition is incorrect, it's all valid, but the individual entries can differ a bit across different libraries/environments.
A similar thing was observed for `inverse` https://github.com/pytorch/pytorch/pull/45034, where the results of cuSOLVER and MAGMA differ a lot for fp32.

mruberry replied:
I think your analysis is probably correct. Would you verify what dtype NumPy computes in and a comment here explaining the absolute tolerance value? 

How significant a loss of precision is 1e-2 here? In terms of relative tolerance, I mean?

mruberry commented on test/test_linalg.py:
```diff
@@ -922,6 +923,90 @@ def test_nuclear_norm_exceptions_old(self, device):
+            actual_L = torch.linalg.cholesky(A)
+            self.assertEqual(actual_L, expected_L)
+
+        shapes = (0, 3, 5)
```
I really appreciate your enumeration of these test cases and testing for degenerate values.

mruberry commented on test/test_linalg.py:
```diff
@@ -922,6 +923,90 @@ def test_nuclear_norm_exceptions_old(self, device):
+        batches = ((), (3, ), (2, 2))
+        larger_input_case = [(100, (5, ))]
+        for shape, batch in list(itertools.product(shapes, batches)) + larger_input_case:
+            run_test(shape, batch)
```
Would you add the complex dtypes for CUDA and assert that an error is thrown. Something like: 

```
if self.device_type == 'cuda' and dtype.is_complex:
  with self.assertRaises(RuntimeError):
    run_test(shape, batch)
else:
   run_test(shape, batch)
```

mruberry replied:
Nevermind. I see you have a separate test for this below. That's totally fine.

mruberry commented on test/test_linalg.py:
```diff
@@ -922,6 +923,90 @@ def test_nuclear_norm_exceptions_old(self, device):
+        with self.assertRaisesRegex(np.linalg.LinAlgError, r'Last 2 dimensions of the array must be square'):
+            np.linalg.cholesky(A.cpu().numpy())
+
+        # cholesky requires the input to be a matrix
```
I think this part of the test can be separated and put into a new test. Maybe something like "test_cholesky_errors"?

mruberry commented on test/test_linalg.py:
```diff
@@ -922,6 +923,90 @@ def test_nuclear_norm_exceptions_old(self, device):
+        with self.assertRaisesRegex(np.linalg.LinAlgError, r'Matrix is not positive definite'):
+            np.linalg.cholesky(A.cpu().numpy())
+
+    # TODO: once there is more support for complex dtypes on GPU, they shall be added to above test
```
Verifying that the complex dtypes throw an error on CUDA will make it so this test MUST be updated when that support is added.

mruberry replied:
Nevermind. I see you have a separate test below.

mruberry commented on test/test_linalg.py:
```diff
@@ -922,6 +923,90 @@ def test_nuclear_norm_exceptions_old(self, device):
+        self.assertEqual(actual_L, expected_L)
+
+    # TODO: enable CUDA tests once
+    # RuntimeError: "triangular_solve_cuda" not implemented for 'ComplexDouble' is fixed
```
This should assert the test fails with CUDA, too.

mruberry commented on test/test_linalg.py:
```diff
@@ -922,6 +923,90 @@ def test_nuclear_norm_exceptions_old(self, device):
+            root = root + torch.eye(shape[-1], dtype=dtype, device=device)
+
+            gradcheck(func, root)
+            # TODO: gradgradcheck does not work correctly yet for complex
```
@anjali411 has been working on better gradcheck support for complex, I'm not sure about the gradgradcheck status. HOWEVER, I think @peterbell10 found a nice workaround for this issue. @peterbell10, can you comment?

IvanYashchuk replied:
`gradgradcheck` works with #45737. Once it's is merged, I'll update the code here.

mruberry replied:
Sounds good.

mruberry commented on test/test_torch.py:
```diff
@@ -7820,19 +7820,9 @@ def cholesky_test_helper(n, batch_dims, upper):
     @dtypes(torch.float32, torch.float64, torch.complex64, torch.complex128)
     @tf32_on_and_off(0.01)
     def test_cholesky(self, device, dtype):
```
Would you move this test to test_linalg.py? It will need a new name, like "test_torch_cholesky" or "test_old_cholesky."

mruberry replied:
Also, is there a change to this test that would demonstrate your lda fix?

Also also, if you spot other Cholesky tests let's move them, too.

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -8,6 +8,51 @@
+cholesky = _add_docstr(_linalg.linalg_cholesky, r"""
+linalg.cholesky(input) -> Tensor
+
+Returns the Cholesky decomposition.
```
I think combining the first few paragraphs would be helpful. What about something like:

"Computes the Cholesky decomposition of a Hermitian positive-definite matrix or the Cholesky decompositions of a batch of such matrices. If the matrices are real-valued then each of their Cholesky decompositions can be written as A = LL^t, for a lower triangular matrix L, and this function returns the matrix L for each input matrix. If the matrices are complex-valued then their Cholesky decompositions are A = L @ L.H for a lower triangular matrix L, where L.H is the conjugate transpose of L."

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -8,6 +8,51 @@
+matrices, then the returned tensor will be composed of lower-triangular Cholesky factors
+of each of the individual matrices.
+
+.. note:: If the :attr:`input` is not Hermitian positive-definite matrix a RuntimeError is raised
```
This note also needs to account for the batching behavior. Maybe something like:

"If :attr:\`input\` is not a Hermitian positive-definite matrix, or if it's a batch of matrices and one of them is not a Hermitian positive-definite matrix, then a RuntimeError will be thrown."

As for the error behavior: does it specify which matrix in the batch was discovered to not be a Hermitian positive-definite matrix, too? Another option here would be to not elaborate about the content of the error message.

IvanYashchuk replied:
Batch is also specified in the error message.
https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/LinearAlgebraUtils.h#L104

mruberry replied:
Great; let's be clear about that (and maybe even add a test for it erroring on the correct batch element).

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -8,6 +8,51 @@
+.. note:: If the :attr:`input` is not Hermitian positive-definite matrix a RuntimeError is raised
+          saying that the input is singular and mentioning which minor of the input matrix is not positive-definite.
+
+.. note::
```
Don't make this a note, it can go write after the preamble (or after a separate mathematical expression if the doc has one). "Supports real inputs and, on the CPU only, complex inputs."

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -8,6 +8,51 @@
+    Backpropagation for complex inputs is only supported on the CPU.
+
+Args:
+    input (Tensor): the input tensor :math:`A` of size :math:`(*, n, n)` where `*` is zero or more
```
Calling the tensor "A" here isn't helpful and possibly misleading if input is a batch of matrices. Maybe you can say, "a Hermitian positive-definite matrix or batch of such matrices."? 

If multiple matrices are passed as input, does each of them have to have the same size? That requirement (or lack thereof) should be mentioned somewhere in the documents.

IvanYashchuk replied:
I'll remove the "A".
The input tensor must have the shape of the form `(*, n, n)`, the last two dimensions are equal, i.e. batch of square matrices. There is a check for that and tests. I could add that matrices should be square, but that's already implicit by calling a matrix positive-definite.

IvanYashchuk replied:
And it's not technically possible to pass multiple matrices as the input of different size, as we don't support list of tensors and consider the last two dimensions of a tensor as a matrix we process.

mruberry replied:
Right. Good point. I've been in too many discussions about nested tensors, which can be ragged, recently. My mistake.

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -8,6 +8,51 @@
+Example::
+
+    >>> a = torch.randn(2, 2, dtype=torch.complex128)
+    >>> a = torch.mm(a, a.t().conj())  # To make a Hermitian
```
Comments should be descriptive, not declarative: "makes a Hermitian positive-definite matrix"

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -8,6 +8,51 @@
+
+    >>> a = torch.randn(2, 2, dtype=torch.complex128)
+    >>> a = torch.mm(a, a.t().conj())  # To make a Hermitian
+    >>> l = torch.linalg.cholesky(a)
```
This is a good example. It'd also be helpful to show a batched example.

mruberry commented on torch/testing/_internal/common_utils.py:
```diff
@@ -1528,6 +1528,14 @@ def random_symmetric_pd_matrix(matrix_size, *batch_dims, **kwargs):
         + torch.eye(matrix_size, dtype=dtype, device=device) * 1e-5
 
 
+def random_hermitian_pd_matrix(matrix_size, *batch_dims, **kwargs):
```
Add a comment describing what this function does and how to call it.

mruberry replied:
I think the signature of this function should be: `(matrix_size, *, batch_dims, dtype=None, device='cpu')`

IvanYashchuk replied:
Okay, I can change that. I was just following the other existing functions.
```python
def random_symmetric_pd_matrix(matrix_size, *batch_dims, **kwargs):
    dtype = kwargs.get('dtype', torch.double)
    device = kwargs.get('device', 'cpu')
    A = torch.randn(*(batch_dims + (matrix_size, matrix_size)),
                    dtype=dtype, device=device)
    return torch.matmul(A, A.transpose(-2, -1))
```
In fact, all calls to `random_symmetric_pd_matrix` can be safely replaced by `random_hermitian_pd_matrix`, probably in a follow-up PR.

mruberry reviewed (commented):
Hey @IvanYashchuk, thanks for this PR!

This PR looks good. I appreciate the lda fix, I like the tests, and the docs are a good start. I made a few test and doc suggestions and have a question about supporting the out argument. 

Let me know if you have any questions and ping me when it's ready for another (likely brief) review. 

Update: test failures are real.

mruberry commented:
Forgot to mention that I think this also needs an entry in overrides.py. I actually thought we had a test that would verify that. Maybe it's not being hit because of the other test failure? 

cc @hameerabbasi

mruberry commented:
Follow-up question, how does the performance of this cholesky compare with `torch.cholesky` and NumPy's cholesky? Is the GPU significantly slower or faster than the CPU? If it is, how would you describe when that happens?

I don't want to take a lot of your time benchmarking. But maybe a small and large single input, then a few batches vs many batches would be interesting to review?

mruberry commented:
> > Forgot to mention that I think this also needs an entry in overrides.py. I actually thought we had a test that would verify that. Maybe it's not being hit because of the other test failure?
> 
> @mruberry We have a whitelist of modules that we check for overrides on in
> 
> https://github.com/pytorch/pytorch/blob/bed3b40523d49dc783c909c4c41c6ed7e12cd62a/torch/overrides.py#L1111-L1116
> 
> It seems `torch.linalg` is not tested.

Interesting. Should we add the torch.linalg module to that list? It may require adding an override or two, but it could save us having to add many overrides later.

Looks like we're also neglecting the new fft module. Should we add a task to the fft tracker to add the overrides?

hameerabbasi commented:
> Interesting. Should we add the torch.linalg module to that list? It may require adding an override or two, but it could save us having to add many overrides later.
> 
> Looks like we're also neglecting the new fft module. Should we add a task to the fft tracker to add the overrides?

Yes, it seems to me we should do both.

mruberry commented:
> Yes, it seems to me we should do both.

Sounds good. I'll update both trackers with a task.

IvanYashchuk committed:
Added the out= variant

IvanYashchuk committed:
Moved error checks to a separate test

IvanYashchuk committed:
Added xfailed test for cholesky cuda autograd

IvanYashchuk committed:
Only complex128 is needed for autograd test

IvanYashchuk committed:
Added a docstring for random_hermitian_pd_matrix

IvanYashchuk committed:
Updated linalg.cholesky docs

IvanYashchuk committed:
Added a note on error message for batch of singular matrices

IvanYashchuk committed:
In tests compare norms of the resulting matrices

IvanYashchuk committed:
Added entry in overrides.py

IvanYashchuk committed:
Added test case for batch singular input

IvanYashchuk committed:
Moved tests for torch.cholesky to test_linalg.py

IvanYashchuk commented:
Here is the performance comparison. The first column is the shape of the input.
I've also included CuPy as its `cupy.linalg.cholesky` is based on cuSOLVER, while PyTorch uses MAGMA currently. As we see in this small experiment cuSOLVER is much faster.
`torch.linalg` column refers to `torch.linalg.cholesky`.
```
[-------------------------------------------- cholesky torch.float64 -------------------------------------------]
                         |  torch.cholesky CUDA  |  torch.linalg CUDA  |   cupy   |  torch.linalg CPU  |   NumPy 
8 threads: ------------------------------------------------------------------------------------------------------
      (3, 3)             |         5106.4        |        3616.0       |    68.9  |        100.4       |      6.1
      (3, 3, 3)          |          731.4        |         731.0       |   311.6  |        110.5       |      6.7
      (3, 3, 3, 3, 3)    |          649.0        |         620.1       |   191.4  |        189.5       |     12.5
      (64, 64)           |         3601.5        |        3699.8       |    97.7  |        183.5       |    219.5
      (3, 64, 64)        |          957.7        |         870.7       |   150.8  |       1510.5       |   2131.6
      (3, 3, 3, 64, 64)  |         1000.7        |         877.6       |   190.1  |       3633.2       |  20863.0
      (512, 512)         |         6748.5        |        6664.3       |  1317.5  |       5377.3       |  40916.4


[------------------------------------------- cholesky torch.float32 -------------------------------------------]
                         |  torch.cholesky CUDA  |  torch.linalg CUDA  |   cupy  |  torch.linalg CPU  |   NumPy 
8 threads: -----------------------------------------------------------------------------------------------------
      (3, 3)             |         5036.4        |        3649.0       |   69.1  |         99.7       |      7.0
      (3, 3, 3)          |          580.5        |         580.5       |  120.2  |        109.4       |      7.7
      (3, 3, 3, 3, 3)    |          920.3        |         915.4       |  272.1  |        239.5       |     26.2
      (64, 64)           |         4951.7        |        4936.6       |   87.7  |        180.8       |   1078.4
      (3, 64, 64)        |          710.1        |         836.1       |  155.0  |       1373.0       |   1033.3
      (3, 3, 3, 64, 64)  |          778.5        |         583.7       |  190.4  |       3536.6       |  23281.6
      (512, 512)         |         4914.6        |        5003.3       |  506.5  |       3917.8       |  30472.2

Times are in microseconds (us).
```

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into linalg-cholesky

mruberry commented:
Thanks for the perf review, @IvanYashchuk. Good idea to include CuPy. Not for this PR, but seems like we can add a follow-up task to the tracker to more carefully review MAGMA vs. cuSolver performance for Cholesky and consider switching to CuPy?

For CPU performance it look like we have a more scalable algorithm but with higher overhead.

IvanYashchuk committed:
Added a dispatch section with DefaultBackend in native_functions.yaml

IvanYashchuk commented:
@mruberry I've updated the documentation.
I've added `out=` variant and tests for that.
Moved test cases that are supposed to raise an error to a separate test.
I've also moved tests for `torch.cholesky` from `test_torch.py`, `test_autograd.py` to `test_linalg.py`, renamed them to `old_cholesky`, and added a test case with empty input (for `lda` parameter fix).

I've refactored the main test a bit. Now if the individual entries of torch and numpy result are too distinct, we compare their matrix norms (with default tolerances) and individual entries with higher tolerance, if matrices are similar in norm then the test is passed.

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into linalg-cholesky

IvanYashchuk committed:
gradgradcheck for cholesky now works

mruberry commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -8814,6 +8814,19 @@
+  python_module: linalg
+  use_c10_dispatcher: full
+  variants: function
+  dispatch:
```
Can we get rid of the dispatch keys here and in the out variant?

IvanYashchuk replied:
It seems that with recent changes this is now a required thing to do.
https://github.com/pytorch/pytorch/tree/master/aten/src/ATen/native#choosing-the-right-dispatch-keyword
> Yes, but you still want to provide a numerically stable gradient formula instead of using autograd, write
dispatch:
  DefaultBackend: kernel

Otherwise, it hits [these line](https://github.com/pytorch/pytorch/blob/master/tools/autograd/gen_variable_type.py#L723-L728). Maybe I'm using it wrong.

mruberry replied:
My mistake. I think you're right and this recently changed.

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -8,6 +8,76 @@
+
+Computes the Cholesky decomposition of a Hermitian (or symmetric for real-valued matrices)
+positive-definite matrix or the Cholesky decompositions for a batch of such matrices.
+The returned matrix ``L`` is lower-triangular, and
```
This needs to be a little clearer about the batching behavior. I think a very small tweak and a reordering should do it:

Each decomposition has the form:

math goes here

where L is a lower-triangular matrix and ...

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -8,6 +8,76 @@
+where :math:`L^H` is the conjugate transpose of :math:`L`, which is just a transpose for the case
+of real-valued input matrices.
+In code it translates to ``input = L @ L.t()` if :attr:`input` is real-valued and
+``input = L @ L.conj().t()`` if :attr:`input` is complex-valued.
```
Then, following the above change, the return value has not yet been mentioned, so a sentence can be added here:

"The :math:`L` matrix from each decomposition is returned."

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -8,6 +8,76 @@
+
+.. note:: If :attr:`input` is not a Hermitian positive-definite matrix, or if it's a batch of matrices
+          and one or more of them is not a Hermitian positive-definite matrix, then a RuntimeError will be thrown.
+          The error message tells the index of the first problematic matrix in the batch.
```
Suggestion:

"If :attr:\`input\` is a batch of matrices, then the error message will include the batch index of the first matrix that is not Hermitian or not positive-definite."

This is a little clearer about the batching behavior and replaces the ambiguous term "problematic."

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -8,6 +8,76 @@
+
+Args:
+    input (Tensor): the input tensor of size :math:`(*, n, n)` where `*` is zero or more
+                batch dimensions consisting of Hermitian positive-definite matrices.
```
It's not that the batch dimensions consistent of Hermitian positive-definite matrices, it's that each matrix in the batch is expected to be Hermitian positive-definite.

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -8,6 +8,76 @@
+            [1.9586+2.0626j, 9.4160+0.0000j]], dtype=torch.complex128)
+
+    >>> a = torch.randn(3, 2, 2, dtype=torch.float64)
+    >>> a = torch.matmul(a, a.transpose(-2, -1))  # makes a symmetric positive-definite matrix
```
Creates a batch of symmetric positive-definite matrices.

mruberry commented on torch/testing/_internal/common_utils.py:
```diff
@@ -1528,6 +1528,19 @@ def random_symmetric_pd_matrix(matrix_size, *batch_dims, **kwargs):
 
 
+def random_hermitian_pd_matrix(matrix_size, *batch_dims, dtype, device):
+    """
```
Excellent docstring.

mruberry commented on test/test_linalg.py:
```diff
@@ -922,6 +924,256 @@ def test_nuclear_norm_exceptions_old(self, device):
+            actual_L = torch.linalg.cholesky(A)
+            matrices_are_equal = np.allclose(actual_L.cpu().numpy(), expected_L)
+
+            # For fp32 individual entries in matrices can differ between PyTorch and NumPy
```
I'm worried about this approach. I think it's very clever but the discrepancy is concerning. 

Is this difference because NumPy is always computing the decomposition in double? Instead of this, should we compare in float with a lower precision, instead?

IvanYashchuk replied:
Yes, NumPy is computing the result in double, then converts it to fp32.
https://github.com/numpy/numpy/blob/v1.19.0/numpy/linalg/linalg.py#L136

So the current approach is: try comparing individual entries with standard tolerances, if it fails then compare the matrices in norm and individual entries with a higher tolerance. Comparing matrices in norm gives us extra confidence that we get something approximately equal as matrix objects, even though the used tolerance (`atol=1e-2, rtol=1e-5`) for comparing individual values in the matrices may seem too high.

mruberry replied:
OK. I'm convinced. That sounds good. Let's just switch NumPy's allclose for PyTorch's testcase's `assertEqual`.

mruberry commented on test/test_linalg.py:
```diff
@@ -922,6 +924,256 @@ def test_nuclear_norm_exceptions_old(self, device):
+        A = random_hermitian_pd_matrix(3, 3, dtype=dtype, device=device)
+        out = torch.empty_like(A)
+        ans = torch.linalg.cholesky(A, out=out)
+        self.assertEqual(ans, out)
```
To be extra safe, add an assertEqual for out and the result of the computation with no out argument set, too.

mruberry commented on test/test_linalg.py:
```diff
@@ -922,6 +924,256 @@ def test_nuclear_norm_exceptions_old(self, device):
+        with self.assertRaisesRegex(RuntimeError, r'to have same size as tensor for argument'):
+            torch.linalg.cholesky(A, out=out)
+
+        # cholesky requires the input to be a square matrix
```
square matrix OR batch of square matrices

mruberry commented on test/test_linalg.py:
```diff
@@ -922,6 +924,256 @@ def test_nuclear_norm_exceptions_old(self, device):
+        with self.assertRaisesRegex(np.linalg.LinAlgError, r'Last 2 dimensions of the array must be square'):
+            np.linalg.cholesky(A.cpu().numpy())
+
+        # cholesky requires the input to be a matrix
```
matrix OR batch of matrices

mruberry replied:
Alternatively, this could just explain that cholesky requires its input to have at least two dimensions

mruberry commented on test/test_linalg.py:
```diff
@@ -922,6 +924,256 @@ def test_nuclear_norm_exceptions_old(self, device):
+        from torch.testing._internal.common_utils import random_hermitian_pd_matrix
+
+        def run_test(shape, batch):
+            A = random_hermitian_pd_matrix(shape, *batch, dtype=dtype, device=device)
```
This is great. Let's add a discontiguous case, too. Maybe the test can be run once with A like this and again where the last two dimensions of A have been swapped?

mruberry commented on test/test_linalg.py:
```diff
@@ -922,6 +924,256 @@ def test_nuclear_norm_exceptions_old(self, device):
+
+    @onlyCPU
+    @skipCPUIfNoLapack
+    @dtypes(torch.float64, torch.complex128)
```
You're only running this test on the CPU, but the next test only verifies that complex autograd fails on CUDA. I think what you want to do is:

```
@skipCUDAIfNoMagma
@skipCPUIfNoLapack
@dtypesIfCPU(torch.float64, torch.complex128)
@dtypes(torch.float64)
```

This will cause complex128 to only be run on the CPU.

mruberry commented:
Hey @IvanYashchuk, I took another look. This is clearly progressing really well. I asked a few questions and made a few suggestions. Looking forward to hearing your thoughts!

IvanYashchuk committed:
Updated documentation for linalg cholesky

IvanYashchuk committed:
Add one more assert for out= test

IvanYashchuk committed:
Updated test_cholesky_errors

IvanYashchuk committed:
Added non contiguous test

IvanYashchuk committed:
Make test_cholesky_autograd run on gpu for fp64

mruberry commented on test/test_linalg.py:
```diff
@@ -922,6 +924,256 @@ def test_nuclear_norm_exceptions_old(self, device):
+                # axis is specified to calculate matrix norm for batched input
+                expected_norm = np.linalg.norm(expected_L, ord=1, axis=(-2, -1))
+                actual_norm = torch.linalg.norm(actual_L, ord=1, axis=(-2, -1))
+                norms_are_equal = np.allclose(actual_norm.cpu().numpy(), expected_norm)
```
Instead of np.allclose this should use self.assertEqual

IvanYashchuk committed:
Changed np.allclose -> torch.allclose

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into linalg-cholesky

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into linalg-cholesky

IvanYashchuk committed:
Fix long lines

IvanYashchuk committed:
Remove unused import

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into linalg-cholesky

rgommers commented:
@IvanYashchuk this PR accumulated a merge conflict

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into linalg-cholesky

IvanYashchuk commented on aten/src/ATen/native/BatchLinearAlgebra.cpp:
```diff
@@ -584,6 +585,25 @@ Tensor& cholesky_out(Tensor &result, const Tensor &self, bool upper) {
+  return at::_cholesky_helper(self, /*upper=*/false).tril_();
+}
+
+Tensor& linalg_cholesky_out(Tensor &result, const Tensor &self) {
```
`at::native::resize_output` should be used here.

IvanYashchuk committed:
Merge branch 'master' into linalg-cholesky

IvanYashchuk committed:
Use at::native::resize_output

IvanYashchuk committed:
Merge branch 'master' into linalg-cholesky

mruberry commented:
Just checking in @IvanYashchuk, do you need anything on this PR? We're really excited to get linalg.cholesky in!

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into linalg-cholesky

IvanYashchuk commented:
Yes, @mruberry, this PR is ready for another review.
Since the last time, I've updated slightly the documentation.
`linalg_cholesky_out` now uses `at::native::resize_output`, and the tests were updated accordingly (a warning is raised if the out tensor doesn't have the correct shape and non-empty and the error is thrown if dtype doesn't match).
Added non-contiguous input test cases.

mruberry commented on aten/src/ATen/native/BatchLinearAlgebra.cpp:
```diff
@@ -535,11 +536,12 @@ static void apply_cholesky(Tensor& self, bool upper, std::vector<int64_t>& infos
   auto self_matrix_stride = matrixStride(self);
   auto batch_size = batchCount(self);
   auto n = self.size(-2);
+  auto lda = std::max(int64_t{1}, n);
```
type the std::max

mruberry commented on aten/src/ATen/native/cuda/BatchLinearAlgebra.cu:
```diff
@@ -1241,10 +1241,11 @@ AT_ERROR("cholesky: MAGMA library not found in "
 
   auto self_data = self.data_ptr<scalar_t>();
   magma_int_t n = magma_int_cast(self.size(-2), "self.size(-2)");
+  auto lda = std::max(magma_int_t{1}, n);
```
type the max

mruberry commented on test/test_linalg.py:
```diff
@@ -1321,6 +1322,278 @@ def test_dot_invalid_args(self, device):
+                self.assertFalse(A.is_contiguous())
+            expected_L = np.linalg.cholesky(A.cpu().numpy())
+            actual_L = torch.linalg.cholesky(A)
+            matrices_are_equal = torch.allclose(actual_L, torch.from_numpy(expected_L).to(device))
```
Use self.assertEqual instead of allclose

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -8,6 +8,77 @@
+          and one or more of them is not a Hermitian positive-definite matrix, then a RuntimeError will be thrown.
+          If :attr:`input` is a batch of matrices, then the error message will include the batch index
+          of the first matrix that is not Hermitian positive-definite.
+
```
Add a warning that this causes cross-device synchronization when called on CUDA inputs

mruberry reviewed (approved):
Nice work, @IvanYashchuk! Just a few small additional comments to cleanup. Let me know when this is ready to be merged.

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into linalg-cholesky

IvanYashchuk committed:
Added a warning about data movement for cuda inputs

IvanYashchuk committed:
Use typed std::max

IvanYashchuk committed:
Replaced torch.allclose with self.assertEqual

IvanYashchuk committed:
Removed unused import

IvanYashchuk commented:
Hey, @mruberry, this is now ready to be merged.

IvanYashchuk committed:
Fix imports

IvanYashchuk commented:
@mruberry, I'm sorry there was an error that I didn't notice when updating the branch. I think this PR needs to be imported again.

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into linalg-cholesky

IvanYashchuk commented:
Resolved merge conflicts.

mruberry commented:
@IvanYashchuk More merge conflicts, unfortunately. You may want to add the tests as a different place in test_linalg.py to try and avoid these conflicts.

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into linalg-cholesky

IvanYashchuk commented:
Merge conflicts are resolved (for now).

IvanYashchuk committed:
Finish merge

IvanYashchuk commented:
@mruberry, I updated the branch.

Commit:
Merge branch 'master' into linalg-cholesky

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into linalg-cholesky

IvanYashchuk committed:
Differentiation of complex cholesky on cuda now works for single input;

batched complex input does not work until batched complex matmul is
implemented

IvanYashchuk committed:
Merge branch 'master' into linalg-cholesky

mruberry commented:
Test failure is real:

test_cholesky_autograd_xfailed_cuda_complex128 - TestLinalgCUDA

IvanYashchuk committed:
Batched matmul for complex on CUDA is implemented now; fix tests

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into linalg-cholesky

IvanYashchuk commented:
So batched matmul for complex-valued inputs on CUDA now works. I fixed the tests.

IvanYashchuk committed:
Remove redundant code from test_old_cholesky_batched

mruberry commented:
This part of the lint failure is real:

```
  {
    path: 'test/test_linalg.py',
    start_line: 214,
    end_line: 214,
    start_column: 57,
    end_column: 57,
    annotation_level: 'failure',
    message: '[E271] multiple spaces after keyword'
  },
```

Sorry for the noise, the other failure is in the base. The ROCm failure appears unrelated.

IvanYashchuk committed:
flake8 fix
